Update Paper Plugins docs (#183)
PaperMC/docs: docs/paper/dev/getting-started/paper-plugins.md

- [Differences](#differences)

## How do I use them?
Similarly to Bukkit plugins, you have to introduce a `paper-plugin.yml` file into your jar resources folder.

> This will not act as a drop in replacement for `plugin.yml`, as some things, as outlined in this guide, need to be declared differently.

Reviewer: ```suggestion
This will not act as a drop-in replacement for `plugin.yml`, as some things, as outlined in this guide, need to be declared differently.
```

> It should be noted you still have the ability to include both `paper-plugin.yml`and `plugin.yml` in the same jar.

Reviewer: ```suggestion
It should be noted you still have the ability to include both `paper-plugin.yml` and `plugin.yml` in the same jar.
```

### Dependency Declaration

Paper Plugins change how to declare dependencies in your `paper-plugin.yml`:

```yml
dependencies:
  bootstrap:

>     # Lets say that RegistryPlugin has some registry elements that this plugin requires.

Reviewer 2: What is a registry element?

Reviewer: Hmm, I'm not really sure how to best describe why a plugin would need to do this *now* without relying on terms from future API. Is there even a use currently to define a bootstrap load order dependency?

Reviewer 3: Correct, that's what I was going for when it comes to that. Although, it's hard since bootstrap still doesn't do anything so it's kind of just all over the place.

    # We don't need this during runtime, so it's not required in the server section. However
    # can be added to both if needed
    RegistryPlugin:
      load: BEFORE
      required: true
      join-classpath: true # Defaults to true
  server:
    # Add a required "RequiredPlugin" dependency, which will load AFTER your plugin.
    RequiredPlugin:
      load: AFTER
      required: true
      # This means that your plugin will not have access to their classpath
      join-classpath: false
```

With Paper Plugins, dependencies are split into two sections:
- `bootstrap` - These are dependencies that you will be using in the [bootstrap](#bootstrapper).
- `server` - These are dependencies that are used for the core functionality of your plugin, whilst the server is running.

> Let's Take a look at a Dependency:

Reviewer 2: ```suggestion
Let's take a look at a dependency:
```

```yml
RegistryPlugin:
  load: BEFORE # Defaults to OMIT
  required: true # Defaults to true
  join-classpath: true # Defaults to true
```

> - `load`: (`BEFORE`|`AFTER`|`OMIT`) Specifies whether this plugin should before or after **your** plugin.

Reviewer 2: What does omit entail?

Author: Might need @Owen1212055's input here

Reviewer: OMIT should mean that the load order will be undefined (or alphabetical). Only BEFORE and AFTER are used to determine the order.

Reviewer 3: It should be marked as undefined then, I don't wanna even promise alphabetical ordering.

- `required`: Whether this plugin is required for your plugin to load.

> - `join-classpath`: Whether your plugin should have access to their classpath.

Reviewer 2: What does having access to a classpath mean/why would you need to change this setting? Even the classpath section at the bottom doesn't *really* make obvious what this setting does

:::note[Cyclic Loading]

> Note that in certain cases plugins may be able to introduce cyclic loading loops, which will prevent the server from starting.

Reviewer 2: ```suggestion
Note that in certain cases, plugins may be able to introduce cyclic loading loops, which will prevent the server from starting.
```

### Dependency Declaration

Paper Plugins change how to declare dependencies in your `paper-plugin.yml`:

```yml
dependencies:
  bootstrap:

>     # Lets say that RegistryPlugin registers some data that your plugin needs to use

Reviewer 4: ```suggestion
    # Let's say that RegistryPlugin registers some data that your plugin needs to use
```

        classpathBuilder.addLibrary(resolver);
    }
}
```
Currently, you are able to add two different library types, `JarLibrary`, and `MavenLibraryResolver`.

## Differences

### Bukkit Serialization System
Paper plugins still support the serialization system (`org.bukkit.configuration.serialization`) that Bukkit uses. However, custom classes will not be
automatically registered for serialization. In order to use `ConfigurationSection#getObject`, 
you **must** call `ConfigurationSerialization.registerClass(Class)` before you attempt to fetch objects from configurations.

### Classloading Isolation

> Paper plugins are not able to access each other unless given explicit access by depending on another plugin, etc. This helps prevent Paper plugins from accidentally accessing each others' dependencies, and in general helps ensure that plugins are only able to access what they explicitly say they're depending on.

Reviewer: ```suggestion
Paper plugins are not able to access each other unless given explicit access by depending on another plugin, etc. This helps prevent Paper plugins from accidentally accessing each other's dependencies, and in general helps ensure that plugins are only able to access what they explicitly depend on.
```

Currently, you are able to add two different library types, ``JarLibrary``, and ``MavenLibraryResolver``.

## Differences

### Bukkit Serialization System
Paper plugins still support the serialization system (``org.bukkit.configuration.serialization``) that Bukkit uses. However, custom classes will not be
automatically registered for serialization. In order to use `ConfigurationSection#getObject`, 
you must call `ConfigurationSerialization.registerClass(Class)` before you attempt to fetch objects from configurations.

### Classloading Isolation
Paper plugins have isolated classloaders, meaning that relocating dependencies will not be necessary.

> Paper plugins are not able to access each other unless given explicit access by depending on another plugin, etc.

Reviewer 2: This is confusing, since it's not a two-way relationship but just one-way by declaring it in your dependency section, and remove the "etc."

This prevents plugins from accidentally accessing your dependencies, and in general helps minimise issues.

You can declare this visibility inside the [dependencies](#dependency-declaration) section:
```yml
Plugin:
  join-classpath: true # Means you have access to their classpath
```

### Load Order Logic Split
In order to better take advantage of classloading isolation, Paper plugins do **not** use the ``dependencies`` field to determine load order.
This was done for a variety of reasons, mostly to allow better control and allow plugins to properly share classloaders.

> See information on [declaring dependencies](#dependency-declaration) for more information on how to declare the load order of your plugin.

Reviewer 3: See ~~information on~~ [declaring dependencies](#dependency-declaration) for more information on how to declare the load order of your plugin.

### Cyclic Plugin Loading

Cyclic loading describes the phenomena when a plugin loading causes a loop which eventually will cycle back to the original plugin.

> Unlike Spigot plugins, Paper plugins will not attempt to resolve cyclic loading issues.

Reviewer 3: bukkit plugins, not spigot plugins
